File: src/main/java/dev/taboo/taboo/util/DateDifference.java
package dev.taboo.taboo.util;

import java.util.ArrayList;

public class DateDifference {

    public static String timeSince(long diff, boolean useSeconds) {
        long secondsInMillis = 1000;
        long minutesInMillis = secondsInMillis * 60;
        long hoursInMillis = minutesInMillis * 60;
        long daysInMillis = hoursInMillis * 24;
        long yearsInMillis = daysInMillis * 365;

        long elapsedYears = diff / yearsInMillis;
        diff = diff % yearsInMillis;

        long elapsedDays = diff / daysInMillis;
        diff = diff % daysInMillis;

        long elapsedHours = diff / hoursInMillis;
        diff = diff % hoursInMillis;

        long elapsedMinutes = diff / minutesInMillis;
        diff = diff % minutesInMillis;

        long elapsedSeconds = diff / secondsInMillis;

        ArrayList<String> output = new ArrayList<>();
        if (elapsedYears > 0)
            output.add(elapsedYears + " year" + (elapsedYears > 1 ? "s" : ""));
        if (elapsedDays > 0)
            output.add(elapsedDays + " day" + (elapsedDays > 1 ? "s" : ""));
        if (elapsedHours > 0)
            output.add(elapsedHours + " hour" + (elapsedHours > 1 ? "s" : ""));
        if (elapsedMinutes > 0)
            output.add(elapsedMinutes + " minute" + (elapsedMinutes > 1 ? "s" : ""));
        if (elapsedSeconds > 0 && useSeconds)
            output.add(elapsedSeconds + " second" + (elapsedSeconds > 1 ? "s" : ""));

        StringBuilder response = new StringBuilder();
        for (var module : output) {
            if (response.toString().equals("")) {
                response = new StringBuilder(module);
            } else if (module.equals(output.get(output.size() - 1))) {
                response.append(", and ").append(module);
            } else {
                response.append(", ").append(module);
            }
        }
        return response.toString();
    }

    public static String timeSince(long diff) {
        return timeSince(diff, true);
    }

}

File: src/main/kotlin/dev/taboo/taboo/util/DateDifference.kt
package dev.taboo.taboo.util

object DateDifference {

    @JvmOverloads
    fun timeSince(diff: Long, useSeconds: Boolean = true): String {
        var difference = diff
        val secondsInMillis: Long = 1000
        val minutesInMillis = secondsInMillis * 60
        val hoursInMillis = minutesInMillis * 60
        val daysInMillis = hoursInMillis * 24
        val yearsInMillis = daysInMillis * 365
        val elapsedYears = difference / yearsInMillis
        difference %= yearsInMillis
        val elapsedDays = difference / daysInMillis
        difference %= daysInMillis
        val elapsedHours = difference / hoursInMillis
        difference %= hoursInMillis
        val elapsedMinutes = difference / minutesInMillis
        difference %= minutesInMillis
        val elapsedSeconds = difference / secondsInMillis
        val output = ArrayList<String>()
        if (elapsedYears > 0) output.add(elapsedYears.toString() + " year" + if (elapsedYears > 1) "s" else "")
        if (elapsedDays > 0) output.add(elapsedDays.toString() + " day" + if (elapsedDays > 1) "s" else "")
        if (elapsedHours > 0) output.add(elapsedHours.toString() + " hour" + if (elapsedHours > 1) "s" else "")
        if (elapsedMinutes > 0) output.add(elapsedMinutes.toString() + " minute" + if (elapsedMinutes > 1) "s" else "")
        if (elapsedSeconds > 0 && useSeconds) output.add(elapsedSeconds.toString() + " second" + if (elapsedSeconds > 1) "s" else "")
        var response = StringBuilder()
        for (module in output) {
            if (response.toString() == "") {
                response = StringBuilder(module)
            } else if (module == output[output.size - 1]) {
                response.append(", and ").append(module)
            } else {
                response.append(", ").append(module)
            }
        }
        return response.toString()
    }

}